### DISCUSSION TITLE
How can I use conda environments to build and install dependencies?

### DISCUSSION THREAD

**Discussion by ianhbell** [usnistgov/Docs4NIST#6] [General]
A key feature of ReadTheDocs is the ability to specify conda/mamba environment to build the docs from. Also, I need to be able to re-compile python extensions as part of the build on the development branch. That works flawlessly on RTD, and I don't see a way to do that in the current implementation. For me that is a dealbreaker.

**Comment by guyer** [comment]
[Apologies, I wasn't "watching" this repo, so missed your question last week]

I have implemented an [experimental branch](https://github.com/usnistgov/NISTtheDocs2Death/tree/conda) that lets you [specify an environment.yml file](https://pages.nist.gov/NISTtheDocs2Death/en/conda/nistthedocs2death.html#conda-environment). For now, to use it, put `uses: usnistgov/NISTtheDocs2Death@conda` in your workflow.  It's all terribly slow right now, because it keeps reinstalling at every step of the overall action; figuring out the [GitHub Container Registry](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry) is already on my todo list.

I've been testing with a [clone of teqp](https://github.com/guyer/teqp/tree/ntd2d). Right now, it's stalled at `doxygen Doxyfile`, evidently because the path isn't what it expects. My script executes `conf.py` in the context of its own directory because that's [what Sphinx does](https://github.com/sphinx-doc/sphinx/blob/master/sphinx/config.py#L353). I'm not sure why `conf.py` would work for you in RtD but not with my code; I took that part directly from Sphinx.

Based on your use of `READTHEDOCS`, it looks like you'll want an `NTD2D` environment variable, or similar. I can provide that.

> I need to be able to re-compile python extensions as part of the build on the development branch

Can you point me to an example of this? I'm not seeing it in teqp.

**Comment by ianhbell** [comment]
Sorry I guess I was thinking about another project where building the extension was part of the doc build. I plan to add that to teqp at some point since currently I can't update the docs until I push wheels to pypi.

On RTD you can also add apt packages, which is how I get doxygen into the RTD instance.  There appears to also be a doxygen package for conda: https://anaconda.org/conda-forge/doxygen

**Reply by guyer** [reply]
It's not that it can't find doxygen; that's installed by your [environment.yml](https://github.com/guyer/teqp/blob/main/doc/environment.yml). The problem is that, when doxygen runs, it errors with [`configuration file Doxyfile not found!`](https://github.com/guyer/teqp/actions/runs/5860598410/job/15889024468#step:4:2656). This appears to be because `cwd=here+'/../..'` doesn't resolve to the base directory of your repo. I haven't had a chance to debug why not.

**Reply by ianhbell** [reply]
How can I debug this on my side?

**Reply by guyer** [reply]
Not sure you need to. I think I figured out something on the plane, but not sure when I'll have a chance to test.

**Reply by guyer** [reply]
I was able to sort out the path issue. (Unsurprisingly) I was not doing exactly what Sphinx does.

**Reply by ianhbell** [reply]
Makes sense, paths and working directories are always tricky in these sorts of problems

**Reply by guyer** [reply]
There are still some gross inefficiencies on my end to sort out (re-re-installing things, mostly; those should be resolved once I understand the Container Registry). Beyond that, my code does not expect importing `conf.py` to be expensive; as such, it loads it a couple of different times in different sub-actions. I can resolve that for teqp by moving `doxygen Doxyfile`, `jupyter nbconvert`, and `sphinx-apidoc` into the `pre-build-command` of my action, but I don't know if I'm missing something in your workflow.

**Comment by guyer** [comment]
I've enabled conda and apt installations, and I refactored to get the runtime down to something semi-reasonable.
Things still aren't error free, but you can see what I've got at usnistgov/teqp#48.

**Reply by ianhbell** [reply]
Where do the docs appear?

**Reply by guyer** [reply]
When the action is running and, if you have appropriate permissions, they'll show up at https://pages.nist.gov/teqp. When the [action runs from my fork](https://github.com/guyer/teqp/actions), it doesn't have permission to push to your `nist-pages` branch (once you have one).

**Comment by ianhbell** [comment]
How can I set up a redirect from the existing docs to the new docs? One last hacked set of docs via RTD?

**Reply by guyer** [reply]
Kamal Choudhary provided me with a bit of JavaScript that can be added to your reST files.
```sphinx
.. raw:: html

    <script type="text/javascript">
    if (String(window.location).indexOf("readthedocs") !== -1) {
        window.alert('The documentation has moved. I will redirect you to the new location.');
        window.location.replace('https://pages.nist.gov/jarvis/');
    }
    </script>
```

I think Kamal put this in all of his reST files, but it seems simpler to me to just create a branch with stub documentation that only serves this JS and then disable all other branches at RtD.

**Reply by ianhbell** [reply]
Yes a branch with the redirect is probably best, build from that branch

**Comment by ianhbell** [comment]
There is something wrong because some notebooks didn't get run with the right teqp version, should be 0.16.0. A handful used older versions, I have no idea how that is possible.

**Reply by guyer** [reply]
It may be getting the old teqp from your [environment.yml](https://github.com/usnistgov/teqp/blob/main/doc/environment.yml) file, which includes teqp in the pip list?

If your package was pure python, you can add 
```python
import pathlib

sys.path.insert(0, pathlib.Path(__file__).parents[2].resolve().as_posix())
```
to your conf.py, but I don't think that'll work here. Does RtD build and install the package?

**Reply by ianhbell** [reply]
RtD just installs wheels, which should be fine since there are manylinux wheels, and those get pulled in for the doc build, which seems to be happening because some notebooks use 0.16, and others don't

**Reply by ianhbell** [reply]
It seems to be because there was already output in the notebook, and the nbconvert skipped them as a result. Is there a force option?

**Reply by guyer** [reply]
Ah, interesting. No `force` option, but it's possible I didn't get the logic right when I migrated [Python in the conf.py](https://github.com/usnistgov/teqp/blob/29fba5863e8af2b6298ce36ed14561610662283b/doc/source/conf.py#L42) to [bash in the workflow](https://github.com/guyer/teqp/blob/7906063c81e48420373733328619ac92c51ca52a/.github/workflows/NISTtheDocs2Death.yml#L17).

**Comment by ianhbell** [comment]
Can you please set up your workflow to be a lot more verbose? It is impossible to debug with no output about what packages it installed and so on

**Reply by guyer** [reply]
Figuring out why there's so little output from the conda install is on my to-do list for this afternoon.
It won't help here, but, in general, you can get stupid amounts of output if you [enable debug logging](https://docs.github.com/en/actions/monitoring-and-troubleshooting-workflows/enabling-debug-logging).

**Reply by guyer** [reply]
`conda env update` doesn't report what it did (and `--verbose` is so noisy that it's useless). I now follow with `conda list`.

**Comment by guyer** [comment]
I've encountered a design issue. One of the first steps I take is to copy the documentation configuration directory before modifying it. I do this because if, e.g., you use [versioneer](https://github.com/python-versioneer/python-versioneer), modifying `docs/` directly will cause the version to show as `.dirty`. This has worked fine for me, but I noticed that once I got teqp working at all with this action, while it ran doxygen, the doxygen output doesn't get linked in the C++ interface section.

I realize now that this is because your Doxygen file says [`OUTPUT_DIRECTORY       = doc/source/_static/doxygen`](https://github.com/usnistgov/teqp/blob/29fba5863e8af2b6298ce36ed14561610662283b/Doxyfile#L71C1-L71C4), which is reasonable enough, but it's not where my `docs-BORGED/` configuration is looking for things.

I don't know offhand how to reconcile this.

**Reply by ianhbell** [reply]
Hmm, I don't either. That's a little bit less important to have working, but still important from a documentation standpoint.
